File: ga/gpt3_5_result_2024_6/result/1968D_gpt/22_1968D.c
fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: i32 = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.trim().split_whitespace();

        let n: i32 = iter.next().unwrap().parse().unwrap();
        let k: i32 = iter.next().unwrap().parse().unwrap();
        let a: i32 = iter.next().unwrap().parse().unwrap();
        let b: i32 = iter.next().unwrap().parse().unwrap();

        let m = std::cmp::min(n, k);
        let mut arr: Vec<i32> = Vec::with_capacity(n as usize);
        let mut aa: Vec<i32> = Vec::with_capacity(n as usize);

        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.trim().split_whitespace();
        for _ in 0..n {
            arr.push(iter.next().unwrap().parse().unwrap());
        }

        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.trim().split_whitespace();
        for _ in 0..n {
            aa.push(iter.next().unwrap().parse().unwrap());
        }

        let mut p1: i64 = 0;
        let mut p2: i64 = 0;
        let mut p1_: i64 = 0;
        let mut p2_: i64 = 0;
        let mut a_: i32 = a;
        let mut b_: i32 = b;

        for x in 1..=m {
            p1 = p1.max(1i64 * aa[a as usize] * (k - x + 1) + p1_);
            p1_ += aa[a as usize];
            a = arr[a as usize];

            p2 = p2.max(1i64 * aa[b as usize] * (k - x + 1) + p2_);
            p2_ += aa[b as usize];
            b = arr[b as usize];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/glm4_result_2024_6/result/1968D_glm/22_1968D.rs
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut lines = input.lines();
    
    let t: usize = lines.next().unwrap().parse().unwrap();
    
    for _ in 0..t {
        let mut parts = lines.next().unwrap().split_whitespace().map(|s| s.parse().unwrap());
        let n: usize = parts.next().unwrap();
        let k: usize = parts.next().unwrap();
        let a: usize = parts.next().unwrap();
        let b: usize = parts.next().unwrap();
        
        let mut arr = vec![0; n + 1];
        let mut aa = vec![0; n + 1];
        
        for i in 1..=n {
            arr[i] = lines.next().unwrap().parse().unwrap();
        }
        
        for i in 1..=n {
            aa[i] = lines.next().unwrap().parse().unwrap();
        }
        
        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;
        let mut a = a;
        let mut b = b;
        
        for _ in 1..=k.min(n) {
            p1 = p1.max(p1_ + aa[a]);
            p1_ += aa[a];
            a = arr[a];
            
            p2 = p2.max(p2_ + aa[b]);
            p2_ += aa[b];
            b = arr[b];
        }
        
        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}